crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,240 @@
-                files.push((message.filename(), message.file.source_text()));
-                index
-            };
+    /// Returns the relative path based on `path` and the package root.
+    ///
+    /// Returns `None` if `path` is not within the package.
+    pub(crate) fn relative_path<'a>(&self, path: &'a Path) -> Option<&'a RelativePath> {
+        path.strip_prefix(&self.package_root).ok()
+    }
 
-            s.serialize_element(&SerializeMessage { message, file_id })?;
+    /// Get the cached results for a single file at relative `path`. This uses
+    /// `file_last_modified` to determine if the results are still accurate
+    /// (i.e. if the file hasn't been modified since the cached run).
+    ///
+    /// This returns `None` if `file_last_modified` differs from the cached
+    /// timestamp or if the cache doesn't contain results for the file.
+    pub(crate) fn get(
+        &self,
+        path: &RelativePath,
+        file_last_modified: SystemTime,
+    ) -> Option<FileCache> {
+        let files = self.files.lock().unwrap();
+        let file = files.get(path)?;
+
+        // Make sure the file hasn't changed since the cached run.
+        if file.last_modified != file_last_modified {
+            return None;
         }
 
-        s.end()
+        Some(file.clone())
+    }
+
+    /// Add or update a file cache at `path` relative to the package root.
+    pub(crate) fn update(&self, path: RelativePathBuf, file: FileCache) {
+        self.files.lock().unwrap().insert(path, file);
     }
-}
 
-struct SerializeMessage<'a> {
-    message: &'a Message,
-    file_id: usize,
+    /// Remove a file cache at `path` relative to the package root.
+    pub(crate) fn remove(&self, path: &RelativePath) {
+        self.files.lock().unwrap().remove(path);
+    }
 }
 
-impl Serialize for SerializeMessage<'_> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let Message {
-            kind,
-            range,
-            fix,
-            // Serialized manually for all files
-            file: _,
-            noqa_offset: noqa_row,
-        } = self.message;
+/// [`Path`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePath = Path;
+/// [`PathBuf`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePathBuf = PathBuf;
 
-        let mut s = serializer.serialize_struct("Message", 5)?;
+/// On disk representation of the cache per source file.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub(crate) struct FileCache {
+    /// Timestamp when the file was last modified before the (cached) check.
+    last_modified: SystemTime,
+    /// Imports made.
+    imports: ImportMap,
+    /// Diagnostic messages.
+    messages: Vec<CacheMessage>,
+    /// Source code of the file.
+    ///
+    /// # Notes
+    ///
+    /// This will be empty if `messages` is empty.
+    source: String,
```

Comment:
My long-term goal for `Message`'s is that they can include code frames pointing to other files. 

For example: You import a private symbol from another file. Ideally, we would create two to three code frames:

1. Code frame showing the import
2. Code frame in the imported file that defines the symbol 
3. (optional) Code frame setting the `__all__` symbol, showing that it does *not* include your import

The challenge with this is that a `Message` now needs access to more than one file. That's why the old implementation used a `Vec` to store the source files.

Reply:
We can still support that I think. We can include a new `code_frames` vector field in `CacheMessage`, which holds a reference to the file and byte range. Maybe that would replace `range`, maybe not, not sure at this time.